### ISSUE TITLE
Default kernel weights initialization of convolution layer

### ISSUE THREAD

**Issue by mjchen611** [pytorch/pytorch#2366]
Default kernel weights initialization of convolution layer. I use the function conv2d, but I can't find the initial  weights of the convolution kernel , or how initialize the weights of convolution kernels?

Can you give me some suggestions ?
Thank you so much.

**Comment by CDLuminate** [comment]
use this module `torch.nn.init`

https://discuss.pytorch.org/t/weight-initilzation/157

**Comment by mjchen611** [comment]
Thank you.

But I am still wondering about the weights and biases initialization about the following code:
this code seems not have obvious initialization operator about weights and biases.

Can you give me some suggestions?

Thank you so much.


from __future__ import print_function
import argparse
import time
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.autograd import Variable
from torch.backends import cudnn

if __name__ == '__main__':
 parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
 parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                     help='input batch size for training (default: 64)')
 parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                     help='input batch size for testing (default: 1000)')
 parser.add_argument('--epochs', type=int, default=10, metavar='N',
                     help='number of epochs to train (default: 10)')
 parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                     help='learning rate (default: 0.01)')
 parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                     help='SGD momentum (default: 0.5)')
 parser.add_argument('--no-cuda', action='store_true', default=False,
                     help='disables CUDA training')
 parser.add_argument('--seed', type=int, default=1, metavar='S',
                     help='random seed (default: 1)')
 parser.add_argument('--log-interval', type=int, default=10000, metavar='N',
                     help='how many batches to wait before logging training status')
 args = parser.parse_args()
 args.cuda = not args.no_cuda and torch.cuda.is_available()

 print('Using CUDA:' + str(args.cuda))

 torch.manual_seed(args.seed)
 if args.cuda:
     torch.cuda.manual_seed(args.seed)

 class Net(nn.Module):
     def __init__(self):
         super(Net, self).__init__()
         self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
         self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
         self.conv2_drop = nn.Dropout2d()
         self.fc1 = nn.Linear(320, 50)
         self.fc2 = nn.Linear(50, 10)


     def forward(self, x):
         x = F.relu(F.max_pool2d(self.conv1(x), 2))  
         x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
         x = x.view(-1, 320)
         x = F.relu(self.fc1(x))
         x = F.dropout(x, training=self.training)
         x = self.fc2(x)
         return F.log_softmax(x)

 model = Net()
 if args.cuda:
     model.cuda()

 # cudnn.enabled = False
 cudnn.benchmark = True

 kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else {}
 train_dataset = datasets.MNIST('../data', train=True, download=True, transform=transforms.Compose([
     transforms.ToTensor(),
     transforms.Normalize((0.1307,), (0.3081,))
 ]))
 test_dataset = datasets.MNIST('../data', train=False, transform=transforms.Compose([
     transforms.ToTensor(),
     transforms.Normalize((0.1307,), (0.3081,))
 ]))
 train_loader = torch.utils.data.DataLoader(
     train_dataset, batch_size=args.batch_size, shuffle=True, **kwargs)
 test_loader = torch.utils.data.DataLoader(
     test_dataset, batch_size=args.batch_size, shuffle=True, **kwargs)

 optimizer = optim.SGD(model.parameters(), lr=args.lr,
                       momentum=args.momentum)

 def train(epoch):
     model.train()

     for batch_idx, (data, target) in enumerate(train_loader):
         if args.cuda:
             data, target = data.cuda(), target.cuda()
         data, target = Variable(data), Variable(target)
         optimizer.zero_grad()
         output = model(data)

         loss = F.nll_loss(output, target)
         loss.backward()
         optimizer.step()

         if batch_idx % args.log_interval == 0:
             print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                       epoch, batch_idx *
                       len(data), len(train_loader.dataset),
                       100. * batch_idx / len(train_loader), loss.data[0]))

 def test(epoch):
     model.eval()
     test_loss = 0
     correct = 0
     for data, target in test_loader:
         if args.cuda:
             data, target = data.cuda(), target.cuda()
         data, target = Variable(data, volatile=True), Variable(target)
         output = model(data)
         test_loss += F.nll_loss(output, target).data[0]
         # get the index of the max log-probability
         pred = output.data.max(1)[1]
         correct += pred.eq(target.data).cpu().sum()

     test_loss = test_loss
     # loss function already averages over batch size
     test_loss /= len(test_loader)
     print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
         test_loss, correct, len(test_loader.dataset),
         100. * correct / len(test_loader.dataset)))

 for epoch in range(1, args.epochs + 1):
     train(epoch)
     test(epoch)

**Comment by CDLuminate** [comment]
Several examples:

https://github.com/pytorch/examples/blob/master/dcgan/main.py#L95-L102

https://github.com/pytorch/vision/blob/master/torchvision/models/vgg.py#L46-L59
